### PR TITLE
8310892: ScopedValue throwing StructureViolationException should be clearer

### PR DESCRIPTION
**Pull request by AlanBateman**
Docs only update to add a missing `@throws StructureViolationException` and make it clearer when the exception is thrown. In the future we might re-visit this so that the description is in one place rather than in each method.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8311024](https://bugs.openjdk.org/browse/JDK-8311024) to be approved

### Issues
 * [JDK-8310892](https://bugs.openjdk.org/browse/JDK-8310892): ScopedValue throwing StructureViolationException should be clearer (**Bug** - P3)
 * [JDK-8311024](https://bugs.openjdk.org/browse/JDK-8311024): ScopedValue throwing StructureViolationException should be clearer (**CSR**)

### Reviewers
 * [Daniel Fuchs](https://openjdk.org/census#dfuchs) (@dfuch - **Reviewer**)
 * [Jaikiran Pai](https://openjdk.org/census#jpai) (@jaikiran - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14679/head:pull/14679` \
`$ git checkout pull/14679`

Update a local copy of the PR: \
`$ git checkout pull/14679` \
`$ git pull https://git.openjdk.org/jdk.git pull/14679/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14679`

View PR using the GUI difftool: \
`$ git pr show -t 14679`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14679.diff">https://git.openjdk.org/jdk/pull/14679.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14679#issuecomment-1609785496)

### CODE REVIEW

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 1]
```diff
@@ -394,19 +394,20 @@ public <T> T get(ScopedValue<T> key) {
...
-         * If {@code op} creates a {@link StructuredTaskScope} but does not {@linkplain
-         * StructuredTaskScope#close() close} it, then exiting {@code op} causes the
-         * underlying construct of each {@code StructuredTaskScope} created in the
-         * dynamic scope to be closed. This may require blocking until all child threads
-         * have completed their sub-tasks. The closing is done in the reverse order that
-         * they were created. Once closed, {@link StructureViolationException} is thrown.
+         * <p> Scoped values are intended to be used in a <em>structured manner</em>. If code
+         * invoked directly or indirectly by the operation creates a {@link StructuredTaskScope}
+         * but does not {@linkplain StructuredTaskScope#close() close} it, then it is detected
+         * as a <em>structure violation</em> when the operation completes (normally or ith an
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:403]:
```suggestion
         * as a <em>structure violation</em> when the operation completes (normally or with an
```

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 2]
```diff
@@ -423,19 +424,20 @@ public <R> R call(Callable<? extends R> op) throws Exception {
...
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:433]:
```suggestion
         * as a <em>structure violation</em> when the operation completes (normally or with an
```

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 3]
```diff
@@ -486,17 +488,18 @@ private <R> R runWith(Snapshot newSnapshot, Callable<R> op) {
...
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:497]:
```suggestion
         * as a <em>structure violation</em> when the operation completes (normally or with an
```

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 4]
```diff
@@ -553,15 +556,15 @@ public static <T> Carrier where(ScopedValue<T> key, T value) {
...
-     * If {@code op} creates a {@link StructuredTaskScope} but does not {@linkplain
-     * StructuredTaskScope#close() close} it, then exiting {@code op} causes the
-     * underlying construct of each {@code StructuredTaskScope} created in the
-     * dynamic scope to be closed. This may require blocking until all child threads
-     * have completed their sub-tasks. The closing is done in the reverse order that
-     * they were created. Once closed, {@link StructureViolationException} is thrown.
+     * <p> Scoped values are intended to be used in a <em>structured manner</em>. If code
+     * invoked directly or indirectly by the operation creates a {@link StructuredTaskScope}
+     * but does not {@linkplain StructuredTaskScope#close() close} it, then it is detected
+     * as a <em>structure violation</em> when the operation completes (normally or ith an
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:565]:
```suggestion
     * as a <em>structure violation</em> when the operation completes (normally or with an
```

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 5]
```diff
@@ -588,15 +592,15 @@ public static <T, R> R callWhere(ScopedValue<T> key,
...
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:601]:
```suggestion
     * as a <em>structure violation</em> when the operation completes (normally or with an
```

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 6]
```diff
@@ -622,15 +627,15 @@ public static <T, R> R getWhere(ScopedValue<T> key,
...
```
Comment (dfuch) [inline src/java.base/share/classes/java/lang/ScopedValue.java:636]:
```suggestion
     * as a <em>structure violation</em> when the operation completes (normally or with an
```

**Review by dfuch — COMMENTED** [review]
I like the new text, it reads well. There's a typo that has been copy pasted several times however.

**Comment by AlanBateman** [comment]
> I like the new text, it reads well. There's a typo that has been copy pasted several times however.

Thanks, I noticed this when looking at the generated javadoc but didn't commit the change to fix. I'll fix it now.

File: src/java.base/share/classes/java/lang/ScopedValue.java [hunk 7]
```diff
@@ -394,19 +394,20 @@ public <T> T get(ScopedValue<T> key) {
...
-         * underlying construct of each {@code StructuredTaskScope} created in the
-         * dynamic scope to be closed. This may require blocking until all child threads
-         * have completed their sub-tasks. The closing is done in the reverse order that
-         * they were created. Once closed, {@link StructureViolationException} is thrown.
+         * <p> Scoped values are intended to be used in a <em>structured manner</em>. If code
+         * invoked directly or indirectly by the operation creates a {@link StructuredTaskScope}
+         * but does not {@linkplain StructuredTaskScope#close() close} it, then it is detected
+         * as a <em>structure violation</em> when the operation completes (normally or with an
+         * exception). In that case, the underlying construct of the {@code StructuredTaskScope}
+         * is closed and {@link StructureViolationException} is thrown.
```
Comment (jaikiran) [inline src/java.base/share/classes/java/lang/ScopedValue.java:405]: Hello Alan, does this mean that the original exception (if any) that caused the completion of the operation would be lost and the caller would instead receive just the `StructureViolationException`?

Comment (AlanBateman) [reply src/java.base/share/classes/java/lang/ScopedValue.java:405]: No, it's just means the SVE is suppressed - see the discussion in the CSR.

Comment (jaikiran) [reply src/java.base/share/classes/java/lang/ScopedValue.java:405]: Thank you, looks good to me.

**Review by jaikiran — APPROVED** [review]

**Review by dfuch — APPROVED** [review]
LGTM

**Pull request closed without merging**